### PR TITLE
vLLM v1 - HPU integration

### PR DESCRIPTION
**Pull request by kzawora-intel**
Follow-up to previous v1 PRs, but with codebase up to date to Feb 14. Lots has changed and the existing older v1 code needs major adjustements too.
Not much is working so far compared to previous versions. This is very much WIP.
- [x] Implemented v1 HPU attn backend, worker, model_runner and executor
- [x] VLLM_USE_V1=1 properly selects V1 HPU components
- [x] V1 HPU executor loads model properly
- [x] V1 HPU executor allocates KV cache properly
- [x] V1 HPU model runner is constructed properly and initializes bucketing
- [x] V1 HPU attention backend gets selected automatically
- [ ] profile_run works on dummy data (not planned in this PR)
- [x] V1 HPU model_runner prepares input tensors based on SchedulerOutputs (rather than SequenceGroupMetadata)
- [x] V1 HPU model_runner differentiates prefill and decode sequences
- [x] V1 HPU model_runner execute_model runs for prefill
- [x] V1 HPU model_runner execute_model runs for decode
- [x] V1 HPU model_runner handles mixed-batch scenarios
- [x] V1 HPU model_runner prefill returns correct results
- [x] V1 HPU model_runner decode returns correct results (w/ flat PA)
- [x] V1 HPU model_runner decode returns correct results (w/ contiguous PA, APC off)
- [x] V1 HPU model_runner prefill runs at BS>1
- [x] V1 standard greedy and random sampling work on HPU
- [x] Capturing and replaying HPU Graphs work
- [ ] V1 HPU model_runner warmup works properly (not planned in this PR)
- [x] V1 HPU automatic prefix caching works properly
- [x] Tensor parallelism works
- [x] torch.compile works
- [x] Llama3.1-8B runs on GSM-8k with SOTA accuracy (APC off)
- [x] Llama3.1-8B runs on GSM-8k with SOTA accuracy (APC on)

### CODE REVIEW

File: vllm/attention/backends/hpu_attn.py [hunk 1]
```diff
@@ -217,16 +217,15 @@ def forward(
...
             key = key.unflatten(0, (block_indices.size(0), -1))
             value = value.unflatten(0, (block_indices.size(0), -1))
+
         if kv_cache is not None and isinstance(kv_cache, tuple):
             key_cache, value_cache = HPUPagedAttention.split_kv_cache(
                 kv_cache, self.num_kv_heads, self.head_size)
 
-            # Reshape the input keys and values and store them in the cache.
-            # If kv_cache is not provided, the new key and value tensors are
-            # not cached. This happens during the initial memory profiling run.
```
Comment (kzawora-intel) [inline vllm/attention/backends/hpu_attn.py:229]: note to self: restore this

File: vllm/entrypoints/llm.py [hunk 2]
```diff
@@ -1416,8 +1416,9 @@ def _run_engine(
         if use_tqdm:
             pbar.close()
 
-        # Make sure that all workers are finished.
-        self.llm_engine.stop_remote_worker_execution_loop()
+        # Make sure that all workers are finished
+        # NOTE(kzawora): this crashes on v1, why?
+        #self.llm_engine.stop_remote_worker_execution_loop()
```
Comment (kzawora-intel) [inline vllm/entrypoints/llm.py:1421]: note to self: check if this is still problematic

File: vllm/v1/sample/sampler.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def forward(
             self.gather_logprobs(raw_logprobs, num_logprobs, token_ids=sampled)
 
         # Use int32 to reduce the tensor size.
-        sampled = sampled.to(torch.int32)
+        sampled = sampled  # .to(torch.int32)
```
Comment (kzawora-intel) [inline vllm/v1/sample/sampler.py:53]: note to self: figure out if we can make this cast functional

File: vllm/v1/worker/hpu_model_runner.py [hunk 4]
```diff
@@ -0,0 +1,2275 @@
...
+        num_blocks = np.ceil(
+            (context_lens + 1) / self.block_size).astype(np.int32).tolist()
+        block_tables_list = []
+        for i, n in enumerate(num_blocks):
+            seq_block_table = block_table_cpu_tensor[i, :n].tolist()
+            assert len(seq_block_table) == n
+            block_tables_list.append(seq_block_table)
+
+        # CONTEXT_LENS [batch_size]
+        #context_lens = (positions.reshape(-1) + 1)
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1426]: noise

File: vllm/v1/worker/hpu_model_runner.py [hunk 5]
```diff
@@ -0,0 +1,2275 @@
...
+        # to compute this.
+        block_number = torch.ones(
+            (padded_batch_size, 1), dtype=torch.int32) * self._PAD_BLOCK_ID
+        block_number[:num_decodes] = torch.gather(input=block_table_cpu_tensor,
+                                                  dim=1,
+                                                  index=(index //
+                                                         self.block_size))
+        block_offsets = padded_index % self.block_size
+        slot_mapping = block_number * self.block_size + block_offsets
+        #import pdb; pdb.set_trace()
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1404]: noise

File: vllm/v1/worker/hpu_model_runner.py [hunk 6]
```diff
@@ -0,0 +1,2275 @@
...
+                               bucketing=True) -> DecodeInputData:
+        # Decodes run as one single padded batch with shape [batch, 1]
+        #
+        # We need to set _PAD_SLOT_ID for the padding tokens in the
+        # slot_mapping, such that the attention KV cache insertion
+        # logic knows to ignore those indicies. Otherwise, the
+        # padding data can be dummy since we have a causal mask.
+
+        block_table_cpu_tensor = self.input_batch.block_table.get_cpu_tensor()
+        # num_reqs = self.input_batch.num_reqs
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1360]: noise

File: vllm/v1/worker/hpu_model_runner.py [hunk 7]
```diff
@@ -0,0 +1,2275 @@
...
+            ]
+            block_groups = [[i] * len(bt)
+                            for i, bt in enumerate(block_tables_list)]
+            block_usage = [
+                [self.block_size] * (len(bt) - 1) + [lbu]
+                for bt, lbu in zip(block_tables_list, last_block_usage) if bt
+            ]
+            req_ids = self.input_batch.req_ids[:num_decodes]
+            self.cache_access_validator.validate_decode(
+                slot_mapping.tolist(), block_tables_list, block_usage, req_ids)
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1469]: remove this

File: vllm/v1/worker/hpu_model_runner.py [hunk 8]
```diff
@@ -0,0 +1,2275 @@
...
+
+        # If num_decodes == self.input_batch.num_reqs, then batch is all decode, and only a single decode forward pass will be executed in this method. # noqa
+        # If num_decodes == 0, then batch is all prefill, and only prefill forward passes will be executed  in this method. # noqa
+        # If neither apply, then batch is mixed, and both prefill and decode forward passes will be executed in this method. # noqa
+
+        # First, we will execute all decodes (if any) in a single batch,
+        # then we'll execute prefills in batches of up to max_prefill_batch_size elements. # noqa
+        # All shapes used in forward passes are bucketed appropriately to mitigate risk of graph recompilations. # noqa
+
+        # We can do sampling directly after executing each forward pass (split_sampler=True), # noqa
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1606]: there's no split_sampler anymore, remove this

File: vllm/v1/worker/hpu_model_runner.py [hunk 9]
```diff
@@ -0,0 +1,2275 @@
...
+        # decode sampler BS4 [D0, D1, D2, 0] -> [tokD0, tokD1, tokD2, 0]
+        # prefill[iter 0] forward pass BS4 [P0, P1, P2, 0]
+        # prefill[iter 0] compute_logits BS4 [P0, P1, P2, 0]
+        # prefill[iter 0] sampler BS4 [P0, P1, P2, 0] -> [tokP0, tokP1, tokP2, 0] # noqa
+        # Join the prefill and decode on device into [tokD0, tokD1, tokD2, 0, tokP0, tokP1, tokP2, 0] # noqa
+        # Transfer [tokD0, tokD1, tokD2, 0, tokP0, tokP1, tokP2, 0] to CPU
+        # On CPU, sanitize [tokD0, tokD1, tokD2, 0, tokP0, tokP1, tokP2, 0] -> [tokD0, tokD1, tokD2, tokP0, tokP1, tokP2] # noqa
+        # Return [tokD0, tokD1, tokD2, tokP0, tokP1, tokP2]
+
+        # Example2: Same thing, but max_prefill_batch_size=4 and split_sampler=False: # noqa
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1645]: this is impossible

File: vllm/v1/worker/hpu_model_runner.py [hunk 10]
```diff
@@ -0,0 +1,2275 @@
...
+                sampling_metadata = self._prepare_sampling(
+                    batch_changed, req_id, pad_to=logits_device.shape[0])
+                sampler_output = self.model.sample(
+                    logits=logits_device, sampling_metadata=sampling_metadata)
+                sampled_token_ids_device = sampler_output.sampled_token_ids
+                htorch.core.mark_step()
+                prefill_output_tokens.extend(sampled_token_ids_device)
+
+            # sampler now returns cpu list instead of device tensor -
+            # and i don't like it
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1700]: it doesn't

File: vllm/v1/worker/hpu_model_runner.py [hunk 11]
```diff
@@ -0,0 +1,2275 @@
...
+            for req_id in self.input_batch.req_ids[:num_reqs]:
+                req_idx = self.input_batch.req_id_to_index[req_id]
+                if self.input_batch.token_ids_cpu[
+                        req_idx, self.input_batch.num_tokens[req_idx] -
+                        1] not in [0, self._tokenizer.eos_token_id, 0]:
+                    continue
+                token_ids = self.input_batch.token_ids_cpu[req_idx]
+                #prompt = self._tokenizer.decode(
+                #    token_ids[:self.input_batch.
+                #              num_prompt_tokens_cpu[req_idx]])
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1799]: noise

File: vllm/v1/worker/hpu_model_runner.py [hunk 12]
```diff
@@ -0,0 +1,2275 @@
...
+
+        model_runner_output = ModelRunnerOutput(
+            req_ids=all_req_ids,
+            req_id_to_index=self.input_batch.req_id_to_index,
+            sampled_token_ids=sampled_token_ids_list,
+            logprobs=None,
+            prompt_logprobs_dict=prompt_logprobs_dict,  # type: ignore[arg-type]
+        )
+
+        if False:
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1789]: debug code, remove this

File: vllm/v1/worker/hpu_model_runner.py [hunk 13]
```diff
@@ -0,0 +1,2275 @@
...
+                req_state = self.requests[req_id]
+                #prompt = self._tokenizer.decode(req_state.prompt_token_ids)
+                generated = self._tokenizer.decode(req_state.output_token_ids)
+                if 'The final answer is' not in generated:
+                    logger.info(
+                        f'[ENGINE_ITER {self._ENGINE_ITER}] REQ:{req_id} IDX:{req_idx} finished: {generated!r}'  # noqa
+                    )
+            logger.info("_ENGINE_ITER %s finished", self._ENGINE_ITER)
+        self._ENGINE_ITER += 1
+        #import pdb; pdb.set_trace()
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1811]: noise

File: vllm/v1/worker/hpu_model_runner.py [hunk 14]
```diff
@@ -0,0 +1,2275 @@
...
+        temperature_device = _async_h2d_tensor_copy(temperature, self.device)
+        top_p_device = _async_h2d_tensor_copy(top_p, self.device)
+        top_k_device = _async_h2d_tensor_copy(top_k, self.device)
+        generators = {
+            i: None
+            for i in range(batch_size)
+        }  # NOTE(kzawora): idk what to set here
+        max_num_logprobs = 0  # NOTE(kzawora): idk what to set here
+        # NOTE(kzawora: do this in a smarter way)
+        return None
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:1935]: yea this aint that smart

File: vllm/v1/worker/hpu_model_runner.py [hunk 15]
```diff
@@ -0,0 +1,2275 @@
...
+                    self.bucketing_ctx.decode_buckets, False, mem_post_decode)
+
+        end_time = time.perf_counter()
+        end_mem = HabanaMemoryProfiler.current_device_memory_usage()
+        elapsed_time = end_time - start_time
+        msg = (
+            f"Warmup finished in {elapsed_time:.0f} secs, "
+            f"allocated {format_bytes(end_mem - start_mem)} of device memory")
+        logger.info(msg)
+        #self.profiler.end()
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_model_runner.py:2169]: remove this

File: vllm/v1/worker/hpu_worker.py [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+        # Reset the seed to ensure that the random state is not affected by
+        # the model initialization and profiling.
+        set_random_seed(self.model_config.seed)
+
+    @torch.inference_mode()
+    def execute_model(
+        self,
+        scheduler_output: "SchedulerOutput",
+    ) -> ModelRunnerOutput:
+        #with track_graph_compile('HPUWorker.execute_model'):
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_worker.py:217]: remove this

File: vllm/v1/worker/hpu_worker.py [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+@contextmanager
+def track_graph_compile(name: str):
+    import habana_frameworks.torch as htorch
+    from habana_frameworks.torch.hpu.metrics import metric_localcontext
+    with metric_localcontext("graph_compilation") as gc:
+        yield
+        htorch.hpu.synchronize()
+    if gc.stats()[0][1] != 0:
+        msg = f"[{name}] graph compilation detected: {gc.stats()}"
+        logger.warning(msg)
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_worker.py:274]: move/remove this? it's pretty nifty tho

File: vllm/v1/worker/hpu_worker.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+    key_cache_block = cache_config.block_size * num_heads * head_size
+    value_cache_block = key_cache_block
+    total = num_attention_layers * (key_cache_block + value_cache_block)
+    if cache_config.cache_dtype == "auto":
+        dtype = model_config.dtype
+    else:
+        dtype = STR_DTYPE_TO_TORCH_DTYPE[cache_config.cache_dtype]
+    dtype_size = get_dtype_size(dtype)
+    return dtype_size * total
```
Comment (kzawora-intel) [inline vllm/v1/worker/hpu_worker.py:262]: this is stale, it'd be best to remove this and its usage

**Review by kzawora-intel — COMMENTED** [review]
smells like fish

File: vllm/v1/worker/hpu_worker.py (hunk 18 above)
Comment (kzawora-intel) [reply vllm/v1/worker/hpu_worker.py]: good job me, you did it!

File: .jenkins/lm-eval-harness/test_lm_eval_correctness.py [hunk 19]
```diff
@@ -51,6 +51,7 @@ def launch_lm_eval(eval_config):
                  f"dtype={dtype}," \
                  f"max_model_len=4096," \
                  f"max_num_seqs={max_num_seqs}," \
+                 f"enable_prefix_caching=False," \
```
Comment (kzawora-intel) [inline .jenkins/lm-eval-harness/test_lm_eval_correctness.py:54]: note: apc is working as well, but it is disabled on v0 by default and enabled on v1 - I wanted to have apples-to-apples comparison and I didn't want to add more complexity for explicitly enabling/disabling it, so I've disabled it accross the board; https://github.com/HabanaAI/vllm-fork/pull/866 is addressing this

File: vllm/v1/worker/gpu_input_batch.py [hunk 20]
```diff
@@ -423,6 +423,75 @@ def make_sampling_metadata(
...
+            presence_penalties=self.presence_penalties[req_indices],
+            repetition_penalties=self.repetition_penalties[req_indices],
+            output_token_ids=output_token_ids,
+            min_tokens=[self.min_tokens[req_idx] for req_idx in req_indices],
+            stop_token_ids=[
+                self.stop_token_ids[req_idx] for req_idx in req_indices
+            ],
+            no_penalties=self.no_penalties,
+        )
+
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/v1/worker/gpu_input_batch.py:494]: (it will, but it's currently the best we've got for preparing SamplingMetadata for arbitrary indices)

File: tests/v1/engine/utils.py [hunk 21]
```diff
@@ -20,7 +20,7 @@
 # Number of prompt logprobs to request when testing prompt logprobs
 NUM_PROMPT_LOGPROBS_UNDER_TEST = 7
 
-TOKENIZER_NAME = "mistralai/Mistral-7B-Instruct-v0.3"
+TOKENIZER_NAME = "/mnt/weka/data/pytorch/mistral/Mistral-7B-Instruct-v0.3"
```
Comment (madamczykhabana) [inline tests/v1/engine/utils.py:23]: wouldn't this cause friction when pulling stuff from upstream?

File: vllm/v1/worker/hpu_model_runner.py [hunk 22]
```diff
@@ -0,0 +1,2118 @@
...
+
+def _async_h2d_tensor_copy(source, device='hpu'):
+    assert source.device.type == 'cpu', \
+        "Source tensor is not present in host memory!"
+    target = torch.empty(source.shape, dtype=source.dtype, device=device)
+    target.copy_(source, non_blocking=True)
+    return target
+
+
+def swap_positions(b: InputBatch, id_1, id_2):
```
Comment (madamczykhabana) [inline vllm/v1/worker/hpu_model_runner.py:196]: This is scary 😨. Could we move this logic to InputBatch? This way this code would be closer to the definition of the class that we're modyfing.

File: tests/v1/engine/utils.py (hunk 21 above)
Comment (kzawora-intel) [reply tests/v1/engine/utils.py:23]: kinda, but not in a significant way, no - if the string gets updated, we'll get a really simple to resolve merge conflict

File: vllm/v1/worker/hpu_model_runner.py (hunk 22 above)
Comment (kzawora-intel) [reply vllm/v1/worker/hpu_model_runner.py:196]: I'd prefer not touching InputBatch if possible - this scary logic is also repeated in tpu_model_runner

**Review by madamczykhabana — APPROVED** [review]
#shipit
